### PR TITLE
Issue #26: Faulty reloading parsing

### PR DESCRIPTION
**Pull request by mdimjasevic**
This PR addresses the following question from issue #26 : What happens if at reloading, one rule (out of several) in one domain (out of several) can't be parsed?

Ratelimit keeps the original rules and this PR adds a test to Fencer to confirm it matches the behavior.

I have two remarks:

* This PR adds a rule reloading test only, and not a server test. A server test would be cumbersome for this as symlinks would have to be manipulated in it and some code from `Fencer.Main.main` would have to be reimplemented. Given that there is no much benefit in such a server test, I added a rule reloading test only.
* The branch is based on the `mdimjasevic/26-cant-read-file` branch. Therefore, please review PR #102 first.

### CODE REVIEW

File: lib/Fencer/Rules.hs [hunk 1]
```diff
@@ -65,14 +65,28 @@ loadRulesFromDirectory
...
+      :: FilePath
+      -> Either Yaml.ParseException DomainDefinition
+         ----------------------------------------------
+      -> Either LoadRulesError (Maybe DomainDefinition)
+    parseErrorHandle _    (Right def)  = Right $ Just def
+    parseErrorHandle file (Left parEx) = case parEx of
+      Yaml.InvalidYaml (Just (Yaml.YamlException _)) ->
+        Right Nothing
+      err ->
+        Left $ LoadRulesParseError file err
```
Comment (effectfully) [inline lib/Fencer/Rules.hs]: That does look really strange that `InvalidYaml` is not considered to be a parse error.

Comment (mdimjasevic) [reply lib/Fencer/Rules.hs]: This was dealt with in PR #102, which this PR is based on. Therefore, this is not in the code anymore.

File: test/Fencer/Rules/Test.hs [hunk 2]
```diff
@@ -227,3 +230,57 @@ test_rulesLoadRulesReadPermissions =
...
+            ("Expected failures, got domains: " ++ show newDefinitions)
+        (Right definitions, f@(Left _)) -> do
+          assertEqual
+            "unexpected failure"
+            (length . toErrorList $ mapLeft NE.toList f)
+            (length . toErrorList $ expectedFailure)
+          assertBool "unexpected definitions"
+            (((==) `on` show)
+             (sortOn domainDefinitionId <$> result)
+             (Right $ sortOn domainDefinitionId definitions))
```
Comment (effectfully) [inline test/Fencer/Rules/Test.hs]: It's a little hard to read these lines. I'd probably use do-notation, but even just

```haskell
          assertBool "unexpected definitions" $
            ((==) `on` show)
              (sortOn domainDefinitionId <$> result)
              (Right $ sortOn domainDefinitionId definitions)
```

is already more readable.

File: test/Fencer/Rules/Test.hs [hunk 3]
```diff
@@ -227,3 +230,57 @@ test_rulesLoadRulesReadPermissions =
...
+            ("Expected domains, got failures: " ++
+             prettyPrintErrors (NE.toList errs))
+        (Right _, Right newDefinitions) ->
+          assertFailure
+            ("Expected failures, got domains: " ++ show newDefinitions)
+        (Right definitions, f@(Left _)) -> do
+          assertEqual
+            "unexpected failure"
+            (length . toErrorList $ mapLeft NE.toList f)
+            (length . toErrorList $ expectedFailure)
```
Comment (effectfully) [inline test/Fencer/Rules/Test.hs]: This is a bit weird. So you have a particular failure

```
"Error in $.descriptors[1]: key \"key\" not present"
```

but then you just remove all the details and only use the fact that there's supposed to be only a single failure. I'd say that you either should write

```haskell
          assertEqual
            "unexpected failure"
            (length . toErrorList $ mapLeft NE.toList f)
            1
```

or compare actual errors, not just the numbers of them.

File: test/Fencer/Rules/Test.hs (hunk 2 above)
Comment (mdimjasevic) [reply test/Fencer/Rules/Test.hs]: That's a fair point! I addressed it with `let`s in an upcoming commit.

File: test/Fencer/Rules/Test.hs (hunk 3 above)
Comment (mdimjasevic) [reply test/Fencer/Rules/Test.hs]: I changed this to comparing errors instead of their count.

File: test/Fencer/Rules/Test.hs (hunk 2 above)
Comment (mdimjasevic) [reply test/Fencer/Rules/Test.hs]: @effectfully: Any remaining issues to be addressed?

File: test/Fencer/Rules/Test.hs [hunk 4]
```diff
@@ -227,3 +229,66 @@ test_rulesLoadRulesReadPermissions =
...
+            ((==) `on` (fmap showError . NE.toList))
+              expectedFailure
+              (trimPath <$> errs)
+ where
+  files :: [RuleFile]
+  files =
+    [ simpleRuleFile ("domain1" </> "config.yml") domainDescriptorKeyValueText
+    , simpleRuleFile ("domain2" </> "config.yml") domainDescriptorKeyText ]
+  files' :: [RuleFile]
+  files' =
```
Comment (effectfully) [inline test/Fencer/Rules/Test.hs]: Nit: I'd name it `filesFailure` or something.

(whenever I say "nit" I mean "feel free to do either way")

**Review by effectfully — APPROVED** [review]
Looks good.

File: test/Fencer/Rules/Test.hs (hunk 4 above)
Comment (mdimjasevic) [reply test/Fencer/Rules/Test.hs]: Done!

**Pull request merged**